src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs

```diff
@@ -547,12 +552,13 @@ private async Task ExecuteApplication(ConnectionDelegate connectionDelegate)
             // Verify some initialization invariants
             Debug.Assert(TransportType != HttpTransportType.None, "Transport has not been initialized yet");
 
-            // Jump onto the thread pool thread so blocking user code doesn't block the setup of the
-            // connection and transport
-            await AwaitableThreadPool.Yield();
+            _connectionDelegate = connectionDelegate;
+
+            // Queue the connection for execution
+            ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
 
-            // Running this in an async method turns sync exceptions into async ones
-            await connectionDelegate(this);
+            // Wait for that task to finish signaling the end of the application execution
+            await _connectionDelegateTcs.Task;
```

Comment:
Why do you choose this route rather than:
```C#
await AwaitableThreadPool.Yield();

Task t;
using (ExecutionContext.SuppressFlow())
    t = connectionDelegate(this);

await t;
```
?

Reply:
I generally don't like using suppress flow. I prefer the more explicit call that doesn't capture the `ExecutionContext`.

Reply:
Also the fact that SuppressFlow throws if the flow is already suppressed makes it annoying. You need to always code around the fact that somebody further up the stack might have already suppressed it.

Reply:
> You need to always code around the fact that somebody further up the stack might have already suppressed it.

There's no user code higher up the stack here.  You shouldn't need to check or need additional code around it.

> I generally don't like using suppress flow

Understood in general.  In practice here, though, the code using SuppressFlow would seem to be less convoluted and (I'd need to measure to know for sure) less expensive.

Reply:
> There's no user code higher up the stack here. You shouldn't need to check or need additional code around it.

There is. That's where the async locals come from and why this change is happening in the first place.

> Understood in general. In practice here, though, the code using SuppressFlow would seem to be less convoluted and (I'd need to measure to know for sure) less expensive.

I think the code does look less convoluted but we use this pattern in other places in ASP.NET Core and it's a stylistic thing. It's the same reason I prefer using *Unsafe APIs instead of suppressing the flow in various cases.

Reply:
> There is. That's where the async locals come from and why this change is happening in the first place.

Does `await AwaitableThreadPool.Yield();` not do what I think it does?

Reply:
Are you saying:

```C#
ExecutionContext.SuppressFlow();

await Task.Yield();

ExecutionContext.SuppressFlow();
```
Works?

Reply:
The other thing I don't like about that is that it's too subtle vs a call that says "This will not capture the execution context" and the code sucks a little bit because. You know what I really want? `Task.UnsafeRun`. That would work here as well.

Reply:
> Are you saying ... works?

Forgetting for the moment the fact that Task.Yield() respects current SynchronizationContext as well (which your AwaitableThreadPool.Yield()) does not, yes... suppression itself doesn't flow, and the thread pool ensures a work item can't leak a change to suppression to the next work item.

Reply:
> You know what I really want? Task.UnsafeRun. That would work here as well.

Why do you want to queue to the ThreadPool? I thought this this change was strictly about the ExecutionContext.

Reply:
I see this change removed the old `await AwaitableThreadPool.Yield();` Never mind me.